clients/src/main/java/org/apache/kafka/common/protocol/types/ArrayOf.java

```diff
@@ -41,6 +41,8 @@ public void write(ByteBuffer buffer, Object o) {
     @Override
     public Object read(ByteBuffer buffer) {
         int size = buffer.getInt();
+        if (size < 0)
+            throw new SchemaException("Array size " + size + " cannot be negative");
```

Comment:
This is a bit of a grey area (we have NULLABLE_BYTES and NULLABLE_STRING for the versions that  accept null) so I kept the existing behaviour but with better error messages. I think we need a discussion on how we want to handle null for the various types in the protocol.